File: leetcode_go/45.Jump_Game_II.go
// https://leetcode.com/problems/jump-game-ii/

func jump(nums []int) int {
    start, end, step := 0,0,0
    if len(nums) == 1 {
        return step
    }
    
    for start <= end {
        new_end := 0
        for i:=start;i<=end;i++ {
            if nums[i]+i > new_end {
                new_end = nums[i]+i
            }
            if new_end >= len(nums)-1 {
                return step +1
            }
        }
        step++
        start = end+1
        end = new_end
    }
    return -1;
}

File: leetcode_java/45.Jump_Game_II.java
// https://leetcode.com/problems/jump-game-ii/

class Solution {
  public int jump(int[] nums) {
    int start = 0;
    int end = 0;
    int step = 0;

    if (nums.length == 1) {
      return 0;
    }

    while (start <= end) {
      int new_end = 0;
      for (int i = start; i <= end; i++) {
        new_end = Math.max(new_end, i + nums[i]);
        if (new_end >= nums.length - 1) {
          return ++step;
        }
      }
      step++;
      start = end + 1;
      end = new_end;
    }

    return -1;
  }
}